text/0000-unsafe-aliased.md

```diff
@@ -0,0 +1,399 @@
+    value: T,
+}
+
+/// When this type is used, that almost certainly means safe APIs need to use pinning
+/// to avoid the aliases from becoming invalidated. Therefore let's mark this as `!Unpin`.
+impl<T> !Unpin for UnsafePinned<T> {}
+
+impl<T: ?Sized> UnsafePinned<T> {
+    /// Constructs a new instance of `UnsafePinned` which will wrap the specified
+    /// value.
+    pub fn new(value: T) -> UnsafePinned<T> where T: Sized {
+        UnsafePinned { value }
+    }
+
+    pub fn into_inner(self) -> T where T: Sized {
+        self.value
+    }
+
+    /// Get read-write access to the contents of an `UnsafePinned`.
+    /// Using `UnsafePinned` without pinning is dangerous,
+    /// so use `get_mut_pinned` instead whenever possible!
+    pub fn get_mut_unchecked(&mut self) -> *mut T {
+        ptr::addr_of_mut!(self.value)
+    }
+
+    /// Get read-write access to the contents of a pinned `UnsafePinned`.
+    pub fn get_mut_pinned(self: Pin<&mut Self>) -> *mut T {
+        // SAFETY: we're not using `get_unchecked_mut` to unpin anything
+        unsafe { ptr::addr_of_mut!(self.get_unchecked_mut().value) }
+    }
+
+    /// Get read-only access to the contents of a shared `UnsafePinned`.
+    /// Note that `&UnsafePinned<T>` is read-only if `&T` is read-only.
+    /// This means that if there is mutation of the `T`, future reads from the
+    /// `*const T` returned here are UB!
+    ///
+    /// ```rust
+    /// unsafe {
+    ///     let mut x = UnsafePinned::new(0);
+    ///     let ref1 = &mut *addr_of_mut!(x);
+    ///     let ref2 = &mut *addr_of_mut!(x);
+    ///     let ptr = ref1.get(); // read-only pointer, assumes immutability
+    ///     ref2.get_mut().write(1);
+    ///     ptr.read(); // UB!
+    /// }
+    /// ```
+    pub fn get(&self) -> *const T {
+        self as *const _ as *const T
+    }
+
+    pub fn raw_get_mut(this: *mut Self) -> *mut T {
+        this as *mut T
+    }
+
+    pub fn raw_get(this: *const Self) -> *const T {
+        this as *const T
+    }
+}
+```
+
+The comment about aliasing `&mut` being "risky" refers to the fact that their safety invariant still asserts exclusive ownership.
+This implies that `duplicate` in the following example, while not causing immediate UB, is still unsound:
+
+```rust
+pub struct S {
+    data: UnsafePinned<i32>,
+}
+
+impl S {
+    fn new(x: i32) -> Self {
+        S { data: UnsafePinned::new(x) }
+    }
+
+    fn duplicate<'a>(s: &'a mut S) -> (&'a mut S, &'a mut S) {
+        let s1 = unsafe { (&s).transmute_copy() };
+        let s2 = s;
+        (s1, s2)
+    }
+}
+```
+
+The unsoundness is easily demonstrated by using safe code to cause UB:
+
+```rust
+let mut s = S::new(42);
+let (s1, s2) = s.duplicate(); // no UB
+mem::swap(s1, s2); // UB
+```
+
+We could even soundly make `get_mut_unchecked` return an `&mut T`, given that the safety invariant is not affected by `UnsafePinned`.
+But that would probably not be useful and only cause confusion.
```

Comment:
This confuses me a bit, I expect to be able to write this kind of function:
```rust
fn safe_ffi_pointer_get() -> &mut UnsafePinned<u32> {
    let ptr: *mut u32 = unsafe { ffi_pointer_get() };
    let ptr: *mut UnsafePinned<u32> = ptr.cast::<UnsafePinned<u32>>();
    let r = unsafe { &mut *ptr };
}
```
But if `get_mut_unchecked` returned `&mut T`, then one could write:
```rust
let r = safe_ffi_pointer_get();
let r = r.get_mut_unchecked();
```

And thus get access to the `&mut u32`, even though it might still be referenced & modified by other pointers from e.g. C.

Reply:
>  even though it might still be referenced & modified by other pointers from e.g. C.

If that might be the case, you *must* use `Pin<&mut UnsafePinned<u32>>` as your return type.

Reply:
I see, I find it a bit unnatural to have to wrap a pointer to `u32` in `Pin`, but it's probably something one will get used to.

Another thing that I noticed was that this essentially gives `Pin<P>` another job. It not only ensures that the pointee is never moved, but also that uniqueness cannot be assumed. This will make writing `unsafe` pin projections much more miserable, since that is an additional requirement one needs to uphold:
```rust
fn project(self: Pin<&mut Self>) -> Pin<&mut Field> {
    // SAFETY: We do not move out of `this` and do not give it to code that might assume uniqueness.
    let this = unsafe { self.get_mut_unchecked() };
    // SAFETY: `self` was pinned and `field` is structually pinned.
    unsafe { Pin::new_unchecked(&mut this.field) }
}
```
I don't think there is another way though. This just makes we want safe field projections even more.
There might be generic code that assumes that `Pin<&mut T>` is actually unique, but if `T` is self referential, that assumption is wrong today, so I don't think that this is a problem.

Reply:
> I see, I find it a bit unnatural to have to wrap a pointer to u32 in Pin, but it's probably something one will get used to.

You explained yourself why it needs to be pinned: "it might still be referenced & modified by other pointers".

> Another thing that I noticed was that this essentially gives Pin<P> another job. It not only ensures that the pointee is never moved, but also that uniqueness cannot be assumed. This will make writing unsafe pin projections much more miserable, since that is an additional requirement one needs to uphold:

I don't follow. This RFC does not introduce any requirements for `Pin` that do not already exist.

Reply:
> > even though it might still be referenced & modified by other pointers from e.g. C.
> 
> If that might be the case, you should use `Pin<&mut UnsafePinned<u32>>` as your return type.

what happens if I do this:
```rust
#[repr(transparent)]
struct Foo {
    field: UnsafePinned<u32>,
}

impl Unpin for Foo {}

fn safe_ffi_pointer_get() -> Pin<&mut Foo> {
    let ptr: *mut u32 = unsafe { ffi_pointer_get() };
    let ptr: *mut Foo = ptr.cast::<Foo>();
    let r = unsafe { &mut *ptr };
    Pin::new(r)
}

let r = safe_ffi_pointer_get().into_inner();
let r = r.field.get_mut_unchecked();
```

So I think `Unpin` has to be made `unsafe` regardless?

Reply:
I added a point to the "alternatives" section that may or may not help here (I'm just guessing where the disagreement comes from here):

> - Instead of adding a new type that needs to be used as `Pin<&mut UnsafePinned<T>>`, can't we just make `Pin<&mut T>` special?
  The answer is no, because working with `Pin<&mut T>` in unsafe code usually involves getting direct access to the `&mut` and then using it "carefully".
  But being careful is not enough when the compiler makes non-aliasing assumptions! We need to preserve the fact that the `&mut T` may have aliases even after `Pin::get_unchecked_mut` was used and inside `Pin::map_unchecked_mut`.
  In a different universe where pinning is a first-class concept with native support for projections and no need for `get_unchecked_mut`, this might not have been required,   but with pinning being introduced as a library type, there is no (currently known) alternative to `UnsafePinned`.

Reply:
> > Another thing that I noticed was that this essentially gives Pin<P> another job. It not only ensures that the pointee is never moved, but also that uniqueness cannot be assumed. This will make writing unsafe pin projections much more miserable, since that is an additional requirement one needs to uphold:
> 
> I don't follow. This RFC does not introduce any requirements for `Pin` that do not already exist. Uniqueness assumptions are completely unaffected by `Pin`. That is indeed the problem that motivates `UnsafePinned` in the first place.

Here is how I interpret the RFC:
The RFC uses `Pin` to ensure no safe access to `&mut UnsafePinned`, since that would allow using it as a normal `&mut T`. The semantics of `&mut T` do not change, `&mut UnsafePinned` behaves the same as `&mut u32` except for codegen not generating `noalias` (and miri treating it specially).
This means that `Pin<&mut T>` might not be unique if `T` contains a `UnsafePinned`, which means that values inside of `T` may change at any time.

Here is what I believe to be the status quo:
`Pin<&mut T>` is unique in the sense that every modification of the value goes through that pointer. So if no writes through pointers derived from that pointer happen, the value will remain unchanged.

I might be wrong on the status quo, since RFL used the `Unpin` hack to allow foreign pointers to change values at any time (of course still serialized). But I always understood it as some sort of gray zone and not really officially supported.

Reply:
> So I think Unpin has to be made unsafe regardless?

I don't follow, why? Note that this RFC introduces a new unstable `UnsafeUnpin` trait that is used to determine whether `&mut` gets `noalias`. `UnsafeUnpin` is the equivalent to `Freeze` for `UnsafeCell`.

---

> Pin\<&mut T> is unique in the sense that every modification of the value goes through that pointer. So if no writes through pointers derived from that pointer happen, the value will remain unchanged.

No, that is already not the case. Futures mutate borrowed fields through aliases of this pointer. That's why they need `UnsafePinned` to begin with.

I am not entirely sure what you mean by "the semantics of", but if you mean the safety invariant, then it is the case that `Pin<&mut T>` has a very different safety invariant from `&mut T`. Consider that if they had the same safety invariant, it would be sound to transmute between them!

I had a formal look at that in [this blog post](https://www.ralfj.de/blog/2018/04/05/a-formal-look-at-pinning.html).

Reply:
> > So I think Unpin has to be made unsafe regardless?
> 
> I don't follow, why? Note that this RFC introduces a new unstable `UnsafeUnpin` trait that is used to determine whether `&mut` gets `noalias`. `UnsafeUnpin` is the equivalent to `Freeze` for `UnsafeCell`.

I have thought about it more and I think I am wrong. I thought that implementing `Unpin` would be a problem for a library, since it can be done without `unsafe`. But it actually is just the problem of pin projections and there the user still needs some `unsafe` code (or `pin-project`, but that ensures `Unpin` is not implemented).

> > Pin<&mut T> is unique in the sense that every modification of the value goes through that pointer. So if no writes through pointers derived from that pointer happen, the value will remain unchanged.
> 
> No, that is already not the case. Futures mutate borrowed fields through aliases of this pointer. That's why they need `UnsafePinned` to begin with. That's why we had a [soundness issue](https://github.com/rust-lang/rust/issues/63818) with futures that's still not properly resolved (but which this RFC properly resolves).

So `Pin<&mut T>` does not have the uniqueness guarantee that `&mut T` has?

> I am not entirely sure what you mean by "the semantics of", but if you mean the safety invariant, then it is the case that `Pin<&mut T>` has a very different semantics than `&mut T`. Consider that if they had the same safety invariant, it would be sound to transmute between them!

The only occurrence of "the semantics of" in my previous comment was:
> The semantics of `&mut T` do not change, `&mut UnsafePinned` behaves the same as `&mut u32` except for codegen not generating `noalias` (and miri treating it specially).

By which I meant that `&mut T` *always* has the uniqueness guarantee as a safety invariant, even if `T = UnsafePinned<U>`.

I think I understand it a bit better now, to summarize:
- `&mut T` has uniqueness as a safety invariant, even if `T = UnsafePinned<U>`.
- `Pin<&mut T>` does not have uniqueness as a safety invariant if and only if `T: UnsafeUnpin`

Is that correct? Because if yes, then I think this should be documented on the `Pin<P>` and `&mut T` types.

Reply:
> Oh I meant "Pin<&mut T> does have uniqueness as a safety invariant if and only if T: UnsafeUnpin".

Still incorrect, it's not an "if and only if". But "T: UnsafeUnpin" implies that the safety invariant of `Pin<&mut T>` is the same as that of `&mut T`.

> What about types T: Unpin + !UnsafeUnpin? I would expect that those are malformed, since you can go safely from Pin<&mut T> -> &mut T but this can violate the safety invariants of &mut T...

I guess that would have to be a type that syntactically contains `UnsafePinned` but does not actually make use of pinning. It doesn't make a ton of sense. See the poll_fn debacle for the kind of problems one can cause with an `impl Unpin for`; `UnsafeUnpin` is an unsafe trait to prevent exactly that.

Basically `UnsafeUnpin` as a trait has a "safety invariant" that is exactly the same as that of `Unpin`, but then it also additionally has validity effects (direct impact on the operational semantics) via the `noalias` for `&mut`. We should only add `noalias` for types that satisfy the Unpin invariant, but we can't trust the `Unpin` trait so we need a new one. Also, like for `Freeze`, it's unclear whether we want to allow programmers to "hide" an `UnsafePinned` with a manual `impl UnsafeUnpin`.

It's a shame that we are pessimizing *all* `&mut UnsafePinned` when we only really need to do then for pinned pointers, but due to a lack of native `&pin` types we can't really do any better.

Reply:
I've written a bit more about `Unpin + !UnsafeUnpin` in the RFC. Basically, these types are a foot-gun -- someone did `impl<T> Unpin MyType<T>` rather than `impl<T: Unpin> Unpin MyType<T>`. It's not *impossible* that this is sound, but it certainly needs extra scrutiny. However, having such a type be `Unpin + !UnsafeUnpin` is still better than it being just `Unpin`; the known poll_fn-related soundness issues in real code would have all been mitigated by this and only theoretical soundness issues like the one I describe [here](https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484) remain.

We also likely would never have found out about the poll_fn situation without those soundness issues in real code... but that's not a great motivation to make code unsound, IMO.^^

Reply:
> > What about types T: Unpin + !UnsafeUnpin? I would expect that those are malformed, since you can go safely from Pin<&mut T> -> &mut T but this can violate the safety invariants of &mut T...
> 
> I guess that would have to be a type that syntactically contains `UnsafePinned` but does not actually make use of pinning. It doesn't make a ton of sense. See [the poll_fn debacle](https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484) for the kind of problems one can cause with an `impl Unpin for`; `UnsafeUnpin` is an unsafe trait to prevent exactly that.

Yes, `impl Unpin for` is a problem.

> Basically `UnsafeUnpin` as a trait has a "safety invariant" that is exactly the same as that of `Unpin`, but then it also additionally has validity effects (direct impact on the operational semantics) via the `noalias` for `&mut`. We should only add `noalias` for types that satisfy the Unpin invariant, but we can't trust the `Unpin` trait so we need a new one. Also, like for `Freeze`, it's unclear whether we want to allow programmers to "hide" an `UnsafePinned` with a manual `impl UnsafeUnpin`.

Would it be far fetched to make `Unpin: UnsafeUnpin`? That would prevent the existence of types `Unpin + !UnsafeUnpin` which IMO don't have a usecase and are foot-guny. Probably need a crater run to see if this breaks something, or do you know of some issue with this?

> It's a shame that we are pessimizing _all_ `&mut UnsafePinned` when we only really need to do then for pinned pointers, but due to a lack of native `&pin` types we can't really do any better.

IMO this isn't so bad -- at least for RFL, all types that will be `!UnsafeUnpin` will only be used when pinned. It would be pretty unnatural to use a `&mut UnsafePinned` directly.
For futures, I don't know enough, but to me it would seem strange to "use" a future that is not pinned.

---

> I've written a bit more about `Unpin + !UnsafeUnpin` in the RFC. Basically, these types are a foot-gun -- someone did `impl<T> Unpin MyType<T>` rather than `impl<T: Unpin> Unpin MyType<T>`. It's not _impossible_ that this is sound, but it certainly needs extra scrutiny. However, having such a type be `Unpin + !UnsafeUnpin` is still better than it being just `Unpin`; the known poll_fn-related cases of UB in real code would have all been mitigated by this and only issues like the one I describe [here](https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484) remain.

The way I see `T: Unpin + !UnsafeUnpin` is that there is no UB unless you give an aliasing `Pin<&mut T>` (or `&mut T`) to arbitrary (safe) code (e.g. `swap`). I am not sure that it is possible to exhibit UB with solely safe code. Because in order to create aliasing `Pin<&mut T>` one still needs `unsafe`. So I agree that this is far better than the status quo.

If the suggestion above to make `Unpin: UnsafeUnpin`, is there some other way to prevent the existence of types with `Unpin + !UnsafeUnpin`? For example creating a lint?

> We also likely would never have found out about the poll_fn situation without that UB in real code... but that's not a great motivation to make code have UB, IMO.^^

Agreed.

Reply:
> Would it be far fetched to make Unpin: UnsafeUnpin? 

I don't see how that can be done backwards-compatibly.
And if we want do devise a plan for a breaking change over an edition boundary, it should IMO involve replacing `Unpin` by `UnsafeUnpin` wholesale... though we'd have to figure out the status of the `Unpin` bounds then, which are common, in contrast to `Freeze` which currently cannot be used in bounds (but that feature is every now and then requested).

I'd prefer to not tackle that challenge in this RFC.

> The way I see T: Unpin + !UnsafeUnpin is that there is no UB unless you give an aliasing Pin<&mut T> (or &mut T) to arbitrary (safe) code (e.g. swap).

Well, the `poll_fn` situation didn't involve any such aliasing and still the example that moves the future is still UB.

> If the suggestion above to make Unpin: UnsafeUnpin, is there some other way to prevent the existence of types with Unpin + !UnsafeUnpin? For example creating a lint?

That would basically lint against all manual `impl Unpin`. I don't know the async ecosystem very well but this feels rather disruptive?

Reply:
> > The way I see T: Unpin + !UnsafeUnpin is that there is no UB unless you give an aliasing Pin<&mut T> (or &mut T) to arbitrary (safe) code (e.g. swap).
> 
> Well, the `poll_fn` situation didn't involve any such aliasing and still the example that moves the future is still UB.

Oh I overlooked that. Still, having `Unpin + !UnsafeUnpin` types feels weird to me.

> > If the suggestion above to make Unpin: UnsafeUnpin, is there some other way to prevent the existence of types with Unpin + !UnsafeUnpin? For example creating a lint?
> 
> That would basically lint against all manual `impl Unpin`. I don't know the async ecosystem very well but this feels rather disruptive?

Not so sure about that. `pin-project` generates `Unpin` impls. An example from tokio:
```rust
pin_project! {
    /// Stream for the [`take`](super::AsyncReadExt::take) method.
    #[derive(Debug)]
    #[must_use = "streams do nothing unless you `.await` or poll them"]
    #[cfg_attr(docsrs, doc(cfg(feature = "io-util")))]
    pub struct Take<R> {
        #[pin]
        inner: R,
        // Add '_' to avoid conflicts with `limit` method.
        limit_: u64,
    }
}
```
Here the `Take<R>: Unpin` if `R: Unpin` and if `R: Unpin` implies `R: UnsafeUnpin`, then this impl would not trigger the lint.

I just don't think we should allow `T: Unpin + !UnsafeUnpin`.

Reply:
>  if R: Unpin implies R: UnsafeUnpin

But that's the thing, it doesn't.

I don't see a good migration path towards not allowing `T: Unpin + !UnsafeUnpin`. And I don't think we should block having `UnsafePinned` on such a migration path. We surely should block stabilizing `UnsafeUnpin` on figuring out what we want to do here, but even stabilizing `UnsafePinned` does not require stabilizing `UnsafeUnpin`.